File: sdk/android/src/main/java/com/tuya/smart/rnsdk/group/TuyaGroupModule.java
package com.tuya.smart.rnsdk.group;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.tuya.smart.home.sdk.TuyaHomeSdk;
import com.tuya.smart.rnsdk.utils.BridgeUtils;
import com.tuya.smart.rnsdk.utils.ReactParamsCheck;
import com.tuya.smart.sdk.api.IGroupListener;
import com.tuya.smart.sdk.api.ITuyaGroup;
import com.tuya.smart.sdk.enums.TYDevicePublishModeEnum;

import java.util.ArrayList;
import java.util.Arrays;

import javax.annotation.Nonnull;

import static com.tuya.smart.rnsdk.utils.Constant.DEVID;
import static com.tuya.smart.rnsdk.utils.Constant.DEVIDS;
import static com.tuya.smart.rnsdk.utils.Constant.DPS;
import static com.tuya.smart.rnsdk.utils.Constant.GROUPID;
import static com.tuya.smart.rnsdk.utils.Constant.GROUPNAME;
import static com.tuya.smart.rnsdk.utils.Constant.PUBLISHMODEENUM;
import static com.tuya.smart.rnsdk.utils.Constant.coverDTL;
import static com.tuya.smart.rnsdk.utils.Constant.getIResultCallback;

public class TuyaGroupModule extends ReactContextBaseJavaModule {
    public TuyaGroupModule(@Nonnull ReactApplicationContext reactContext) {
        super(reactContext);
    }

    private IGroupListener mIGroupListener;


    @ReactMethod
    public void registerGroupListener(final ReadableMap params) {
        if (ReactParamsCheck.checkParams(Arrays.asList(GROUPID), params)) {
            if (mIGroupListener != null) {
                getITuyaGroup(params.getDouble(GROUPID)).unRegisterGroupListener();
                mIGroupListener = null;
            }
            mIGroupListener = new IGroupListener() {
                @Override
                public void onDpUpdate(long groupId, String dps) {
                    WritableMap map = Arguments.createMap();
                    map.putDouble("groupId", groupId);
                    map.putString("dps", dps);
                    map.putString("type", "onDpUpdate");
                    BridgeUtils.groupListener(getReactApplicationContext(), map, coverDTL(params.getDouble(GROUPID)) +"");
                }

                @Override
                public void onGroupInfoUpdate(long groupId) {
                    WritableMap map = Arguments.createMap();
                    map.putDouble("groupId", groupId);
                    map.putString("type", "onGroupInfoUpdate");
                    BridgeUtils.groupListener(getReactApplicationContext(), map, coverDTL(params.getDouble(GROUPID)) +"");
                }

                @Override
                public void onGroupRemoved(long groupId) {
                    WritableMap map = Arguments.createMap();
                    map.putDouble("groupId", groupId);
                    map.putString("type", "onGroupRemoved");
                    BridgeUtils.groupListener(getReactApplicationContext(), map, coverDTL(params.getDouble(GROUPID)) +"");
                }
            };
            getITuyaGroup(params.getDouble(GROUPID)).registerGroupListener(mIGroupListener);
        }

    }

    @ReactMethod
    public void unRegisterHomeStatusListener(ReadableMap params) {
        if (mIGroupListener != null) {
            getITuyaGroup(params.getDouble(GROUPID)).unRegisterGroupListener();
            mIGroupListener = null;
        }
    }


    @Nonnull
    @Override
    public String getName() {
        return "TuyaGroupModule";
    }


    @ReactMethod
    public void renameGroup(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(GROUPID, GROUPNAME), params)) {
            getITuyaGroup(params.getDouble(GROUPID)).renameGroup(GROUPNAME, getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void dismissGroup(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(GROUPID), params)) {
            getITuyaGroup(params.getDouble(GROUPID)).dismissGroup(getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void publishDps(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(GROUPID, DPS), params)) {
            getITuyaGroup(params.getDouble(GROUPID)).publishDps(params.getString(DPS), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void publishDpsWithEnum(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(GROUPID, DPS, PUBLISHMODEENUM), params)) {
            getITuyaGroup(params.getDouble(GROUPID)).publishDps(params.getString(DPS), TYDevicePublishModeEnum.valueOf(params.getString(PUBLISHMODEENUM)), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void updateDeviceList(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(GROUPID, DEVIDS), params)) {
            ArrayList<String> list = new ArrayList<>();
            for (int i = 0; i < params.getArray(DEVIDS).size(); i++) {
                list.add(params.getArray(DEVIDS).getString(i));
            }
            getITuyaGroup(params.getDouble(GROUPID)).updateDeviceList(list, getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void addDevice(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(GROUPID, DEVID), params)) {
            getITuyaGroup(params.getDouble(GROUPID)).addDevice(params.getString(DEVID), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void removeDevice(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(GROUPID, DEVID), params)) {
            getITuyaGroup(params.getDouble(GROUPID)).removeDevice(params.getString(DEVID), getIResultCallback(promise));
        }
    }

    private ITuyaGroup getITuyaGroup(double groupId) {
        return TuyaHomeSdk.newGroupInstance(coverDTL(groupId));
    }

}

File: android/src/main/java/com/tuya/smart/rnsdk/group/TuyaGroupModule.kt
package com.tuya.smart.rnsdk.group

import com.facebook.react.bridge.*
import com.tuya.smart.home.sdk.TuyaHomeSdk
import com.tuya.smart.home.sdk.api.ITuyaHome
import com.tuya.smart.home.sdk.callback.ITuyaResultCallback
import com.tuya.smart.rnsdk.utils.*
import com.tuya.smart.rnsdk.utils.Constant.COMMAND
import com.tuya.smart.rnsdk.utils.Constant.DEVIDS
import com.tuya.smart.rnsdk.utils.Constant.GROUPID
import com.tuya.smart.rnsdk.utils.Constant.HOMEID
import com.tuya.smart.rnsdk.utils.Constant.NAME
import com.tuya.smart.rnsdk.utils.Constant.PRODUCTID
import com.tuya.smart.rnsdk.utils.Constant.getIResultCallback
import com.tuya.smart.sdk.api.*
import com.tuya.smart.sdk.bean.GroupDeviceBean




class TuyaGroupModule(reactContext: ReactApplicationContext?) : ReactContextBaseJavaModule(reactContext) {
    override fun getName(): String {
        return "TuyaGroupModule"
    }

    /*创建群组*/
    @ReactMethod
    fun createGroup(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(HOMEID, PRODUCTID, NAME, DEVIDS), params)) {
            getITuyaHome(params.getDouble(HOMEID).toLong()).createGroup(
                    params.getString((PRODUCTID)),
                            params.getString(NAME),
                            JsonUtils.parserArraybyMap(params.getArray(DEVIDS),
                                    String::class.java) as MutableList<String>?,
                            object : ITuyaResultCallback<Long> {
                                override fun onSuccess(p0: Long) {
                                    promise.resolve(p0)
                                }

                                override fun onError(p0: String?, p1: String?) {
                                    promise.reject(p0,p1)
                                }
                            }
                    )
        }
    }

    /**
     * 此接口主要是从云端拉取最新群组列表 根据产品ID
     */
    @ReactMethod
    fun queryDeviceListToAddGroup(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(HOMEID,PRODUCTID), params)) {
            getITuyaHome(params.getDouble(HOMEID).toLong()).queryDeviceListToAddGroup(params.getString(PRODUCTID),
                    object : ITuyaResultCallback<List<GroupDeviceBean>>{
                override fun onSuccess(bizResult: List<GroupDeviceBean>) {
                    promise.resolve(TuyaReactUtils.parseToWritableArray(JsonUtils.toJsonArray(bizResult)))
                }

                override fun onError(errorCode: String, errorMsg: String) {
                    promise.reject(errorCode,errorMsg)
                }
            })
        }
    }

    @ReactMethod
    fun dismissGroup(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(GROUPID), params)) {
            getITuyaGroup(params.getDouble(GROUPID).toLong())
                    ?.dismissGroup(getIResultCallback(promise)
                    )
        }
    }

    @ReactMethod
    fun registerGroupListener(params: ReadableMap) {
        if (ReactParamsCheck.checkParams(arrayOf(GROUPID), params)) {
            getITuyaGroup(params.getDouble(GROUPID).toLong())
                    ?.registerGroupListener(object : IGroupListener {
                        override fun onDpUpdate(var1: Long, var3: String){
                            var map=Arguments.createMap();
                            map.putDouble("id",var1.toDouble())
                            map.putString("dps",var3)
                            map.putString("type","onDpUpdate")
                            BridgeUtils.groupListener(reactApplicationContext,map,params.getDouble(GROUPID).toLong())
                        }

                        override fun onGroupInfoUpdate(var1: Long){
                            var map=Arguments.createMap();
                            map.putDouble("id",var1.toDouble())
                            map.putString("type","onGroupInfoUpdate")
                            BridgeUtils.groupListener(reactApplicationContext,map,params.getDouble(GROUPID).toLong())
                        }

                        override fun onGroupRemoved(var1: Long){
                            var map=Arguments.createMap();
                            map.putDouble("id",var1.toDouble())
                            map.putString("type","onGroupRemoved")
                            BridgeUtils.groupListener(reactApplicationContext,map,params.getDouble(GROUPID).toLong())
                        }
                    })
        }
    }

    @ReactMethod
    fun unregisterGroupListener(params: ReadableMap) {
        if (ReactParamsCheck.checkParams(arrayOf(GROUPID), params)) {
            getITuyaGroup(params.getDouble(GROUPID).toLong())
                    ?.unRegisterGroupListener()
        }
    }
    @ReactMethod
    fun publishDps(params: ReadableMap,promise: Promise){
        if (ReactParamsCheck.checkParams(arrayOf(GROUPID,COMMAND), params)) {
            getITuyaGroup(params.getDouble(GROUPID).toLong())
                    ?.publishDps(JsonUtils.toString(TuyaReactUtils.parseToMap(params.getMap(COMMAND))), getIResultCallback(promise))
        }
    }

    @ReactMethod
    fun onDestroy(params: ReadableMap){
        if (ReactParamsCheck.checkParams(arrayOf(GROUPID,COMMAND), params)) {
            getITuyaGroup(params.getDouble(GROUPID).toLong())
                    ?.onDestroy()
        }
    }

    fun getITuyaHome(homeId: Long): ITuyaHome{
        return TuyaHomeSdk.newHomeInstance(homeId)
    }

    fun getITuyaGroup(groupId: Long): ITuyaGroup{
        return TuyaHomeSdk.newGroupInstance(groupId)
    }

}